### PR TITLE
SAK-44170 Profile Pic Editor: make it Consistent & improve Accessibility

### PR DESCRIPTION
**Pull request by V100X**
Replaced the old Profile2 editor in favor of `sakai.morpheus.usernav.js + site.vm` -> Portal + Library.
See advantages: [SAK-44170](https://sakaiproject.atlassian.net/browse/SAK-44170)

Made improvements to the remaining pic editor:

**· Accessibility**

- Now you can open the editor on the right profile menu using only keyboard.

- Added the appropriate labels for screen reader on edit buttons like “zoom” or “rotate”. (will need translation)

- After pressing ESC or Cancel button, focus will return to the element used to open the editor modal.

**· Properties**

The new editor wasn’t taking into account the properties related to profile pic.

      profile2.picture.change.enabled = true / false

      profile2.picture.type = official / upload / url*

Now, both ways to the pic editor will be disabled by the suitable properties.

**BUT**, attach picture by URL isn’t being supported. This feature should be implemented on the new editor through a new Jira.

___

The image on Profile tool and top bar will be instantly updated.

Also made some design improvements on the Profile Tool and Pic Editor.

### CODE REVIEW

File: profile2/impl/src/java/org/sakaiproject/profile2/logic/ProfileImageLogicImpl.java [hunk 1]
```diff
@@ -1129,6 +1133,22 @@ private String getAvatarInitialsColor(String displayName) {
...
+		UserProfile userProfile = (UserProfile) profileLogic.getUserProfile(userUuid);
+		ProfilePreferences prefs = preferencesLogic.getPreferencesRecordForUser(userUuid);
+		// -is picture changing disabled? (property, or locked).
+		// -if using official images, is the user allowed to select an alternate?
+		//  or have they specified the official image in their preferences.
+		// -last condition prevents the editor being enabled when only official img is allowed by property. 
+		if((!sakaiProxy.isProfilePictureChangeEnabled() || userProfile.isLocked()) && !sakaiProxy.isSuperUser() ||
+				sakaiProxy.isOfficialImageEnabledGlobally() && (!sakaiProxy.isUsingOfficialImageButAlternateSelectionEnabled() || prefs.isUseOfficialImage()) ||
+					sakaiProxy.isProfilePictureChangeEnabled() && sakaiProxy.isUsingOfficialImage()) {
+			return false;
```
Comment (V100X) [inline profile2/impl/src/java/org/sakaiproject/profile2/logic/ProfileImageLogicImpl.java:1147]: Moved logic to its own method as now will be used in various places.

Added an extra condition to avoid the following situation:
When `profile2.picture.change.enabled=true` & `profile2.picture.type=official`
Student apparently could change its profile pic, but after clicking _save_, the new img wouldn't apply.

File: library/src/morpheus-master/js/src/sakai.morpheus.usernav.js [hunk 2]
```diff
@@ -1,10 +1,9 @@
-/**
- * ESC handler to dismiss user nav
- */
+let modalVisible = false;
 
+// ESC handler to dismiss user nav
 function userNavEscHandler(e){
 
-  if (e.keyCode === 27) { // esc keycode
+  if (e.keyCode === 27 && modalVisible != true) { // esc keycode
```
Comment (mpellicer) [inline library/src/morpheus-master/js/src/sakai.morpheus.usernav.js:6]: !modalVisible. I'm consufed about this, should be modalVisible?

File: library/src/morpheus-master/js/src/sakai.morpheus.usernav.js [hunk 3]
```diff
@@ -173,19 +175,20 @@ $PBJQ('.Mrphs-userNav__pic-changer').on("click", function (event) {
...
 
-        var picLink = $PBJQ('#loginUser > .Mrphs-userNav__submenuitem--profilepicture');
-        var parent = picLink.parent();
-        picLink.detach();
-        var d = new Date();
-        var style = 'background-image: url(/direct/profile/' + portal.user.id + '/image/thumb?' + d.getTime() + ')';
+        const d = new Date();
+
+        const picProfile = $PBJQ('#myPhoto');
+        if (picProfile != null) {
```
Comment (mpellicer) [inline library/src/morpheus-master/js/src/sakai.morpheus.usernav.js:181]: I think "if (picProfile)" is enough here?

File: library/src/morpheus-master/js/src/sakai.morpheus.usernav.js [hunk 4]
```diff
@@ -173,19 +175,20 @@ $PBJQ('.Mrphs-userNav__pic-changer').on("click", function (event) {
...
+            picProfile.attr("src", "/direct/profile/" + portal.user.id + "/image?" + d.getTime());
```
Comment (mpellicer) [inline library/src/morpheus-master/js/src/sakai.morpheus.usernav.js:182]: Use String templates please:

`/direct/profile/${portal.user.id}/image?${d.getTime()}`;

File: library/src/morpheus-master/js/src/sakai.morpheus.usernav.js [hunk 5]
```diff
@@ -173,19 +175,20 @@ $PBJQ('.Mrphs-userNav__pic-changer').on("click", function (event) {
...
+        }
+
+        const style = 'background-image: url(/direct/profile/' + portal.user.id + '/image/thumb?' + d.getTime() + ')';
```
Comment (mpellicer) [inline library/src/morpheus-master/js/src/sakai.morpheus.usernav.js:185]: Same as above, String templates are better.

**Review by mpellicer — CHANGES_REQUESTED** [review]

**Comment by mpellicer** [comment]
Good job so far @V100X , just some minor implementation comments.

File: profile2/impl/src/java/org/sakaiproject/profile2/logic/ProfileImageLogicImpl.java (hunk 1 above)
Comment (mpellicer) [reply profile2/impl/src/java/org/sakaiproject/profile2/logic/ProfileImageLogicImpl.java:1147]: I find this complex, hard to read and hard to maintain, could we split this in simple conditions?

**Review by mpellicer — CHANGES_REQUESTED** [review]

File: profile2/impl/src/java/org/sakaiproject/profile2/logic/SakaiProxyImpl.java [hunk 6]
```diff
@@ -1273,6 +1273,18 @@ public boolean isOfficialImageEnabledGlobally() {
 				ProfileConstants.SAKAI_PROP_PROFILE2_OFFICIAL_IMAGE_ENABLED);
 	}
 
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public boolean isUsingOfficialImage() {
+
+		if (getProfilePictureType() == ProfileConstants.PICTURE_SETTING_OFFICIAL) {
```
Comment (mpellicer) [inline profile2/impl/src/java/org/sakaiproject/profile2/logic/SakaiProxyImpl.java:1282]: One line here:

return getProfilePictureType() == ProfileConstants.PICTURE_SETTING_OFFICIAL;

**Review by mpellicer — CHANGES_REQUESTED** [review]

File: profile2/tool/src/java/org/sakaiproject/profile2/tool/pages/MyProfile.java [hunk 7]
```diff
@@ -231,76 +230,21 @@ private void renderMyProfile(final String userUuid, final String requestedTab) {
 			userProfile.setLocked(this.isLocked());
 		}
 
-
```
Comment (mpellicer) [inline profile2/tool/src/java/org/sakaiproject/profile2/tool/pages/MyProfile.java:234]: LOVELY RED LINES!

File: library/src/morpheus-master/js/src/sakai.morpheus.usernav.js (hunk 2 above)
Comment (V100X) [reply library/src/morpheus-master/js/src/sakai.morpheus.usernav.js:6]: It was a quick fix. `!modalVisible` condition prevents the profile menu from closing when the Editor modal is active and you press ESC.

This way, once you close the Editor modal, the focus will be able to get back to the initial element, in this case, the avatar img on the profile menu.

focusBehaviourCrop

A better solution might involve remaking the profile menu, perhaps using the BS modal classes.

File: library/src/morpheus-master/sass/modules/tool/profile2/_profile2.scss [hunk 8]
```diff
@@ -60,8 +63,11 @@
 		clip: rect(1px, 1px, 1px, 1px);
 	}
 
-	#myPhoto:hover > .edit-image-button, .edit-image-button:focus {
+	.btn-holder:hover > .edit-image-button,
+	.edit-image-button:focus > .edit-image-button:focus-visible,
+	.edit-image-button:focus-visible {
```
Comment (V100X) [inline library/src/morpheus-master/sass/modules/tool/profile2/_profile2.scss:68]: `:focus-visible` is usually used to show the blue outline over input elements when keyboard is used, and not when using mouse.

In this case is specially useful, as `edit-image-button` in meant to show only on hover, so when the Edit modal is closed, focus gets back to this button through JS and thanks to SCSS is 'converted' to `:focus-visible`.

Therefore, if you're only using mouse, and the cursor isn't hovering the img after you **close the Editor modal**, the button will hide. However, it will remain if you're interacting with keyboard, in order to get to the last used element without needing to hit the tab key X times.

Best behaviour for each scenario : )

**Review by mpellicer — APPROVED** [review]

File: portal/portal-render-engine-impl/impl/src/webapp/vm/morpheus/site.vm [hunk 9]
```diff
@@ -246,25 +246,31 @@
...
                             </div>
-                            <div class="btn-group">
-                                <a class="profile-image-pan-up btn btn-sm btn-default" href="javascript:void(0)"></a>
-                                <a class="profile-image-pan-down btn btn-sm btn-default" href="javascript:void(0)"></a>
-                                <a class="profile-image-pan-left btn btn-sm btn-default" href="javascript:void(0)"></a>
-                                <a class="profile-image-pan-right btn btn-sm btn-default" href="javascript:void(0)"></a>
-                            </div>
-                            <div class="btn-group">
-                                <a class="profile-image-rotate btn btn-sm btn-default" href="javascript:void(0)"></a>
+                            <div id="cropToolbar" class="btn-toolbar">
```
Comment (ottenhoff) [inline portal/portal-render-engine-impl/impl/src/webapp/vm/morpheus/site.vm:259]: id=cropToolbar is used twice

Comment (ottenhoff) [reply portal/portal-render-engine-impl/impl/src/webapp/vm/morpheus/site.vm:259]: https://sakaiproject.atlassian.net/browse/SAK-47135

Comment (V100X) [reply portal/portal-render-engine-impl/impl/src/webapp/vm/morpheus/site.vm:259]: Of course
